toggler/toggler.py

```diff
@@ -0,0 +1,169 @@
+) -> dict:
+    if stream is not None:
+        return yaml.safe_load(stream)
+    if path is not None:
+        with open(path) as f:
+            return yaml.safe_load(f)
+    path_cfg = os.environ.get(ENV_KEY_CFG)
+    if path_cfg:
+        with open(path_cfg) as f:
+            return yaml.safe_load(f)
+    raise ValueError(
+        f"To initiate Toggler, either stream, path or {ENV_KEY_CFG}"
+        + " environ must be specified to load configuration!"
+    )
+
+
+class Toggler:
+    """Class to parse feature flags configuration file for environment."""
+
+    def __init__(
+        self,
+        mode: str | None = None,
+        stream: str | bytes | FileIO | None = None,
+        path: str | pathlib.Path | None = None,
+        check_deadlines: bool = True,
+    ):
+        self._mode = self._prepare_mode(mode)
+        self._cfg = _load_cfg(stream=stream, path=path)
+        self._features_cfg = self._parse_cfg()
+        if check_deadlines:
+            self.check_deadlines()
+
+    @property
+    def features_cfg(self) -> dict[str, dict[str, Feature]]:
+        return self._features_cfg
+
+    @property
+    def mode_features(self) -> dict[str, Feature]:
+        """Return features for current mode."""
+        return self.features_cfg[self.mode]
+
+    @property
+    def mode(self) -> str:
+        return self._mode
+
+    def is_active(self, name: str) -> bool | None:
+        """Check if current mode feature is enabled.
+
+        If feature is not explicitly defined on current mode, None value
+        is returned instead.
+        """
+        # Values coming from environ can temporary override existing
+        # features, to make it more flexible!
+        env_val = self._is_active_via_env(name)
+        if env_val is not None:
+            return env_val
+        feature = self.mode_features.get(name)
+        if feature is None:
+            try:
+                # We now look if such feature exists in any of modes, to
+                # inform about it.
+                self._search_feature(name)
+            except MissingFeatureError as e:
+                _logger.warning(e.args[0])
+            return None
+        return feature.active
+
+    def check_deadlines(self) -> list[Feature]:
+        """Check if any feature is passed deadline.
+
+        Returns:
+            features that have expired.
+
+        """
+        today = datetime.date.today()
+        expired = []
+        for cfg_values in self.features_cfg.values():
+            for feature in cfg_values.values():
+                if feature.deadline and feature.deadline <= today:
+                    logging.warning(
+                        'Feature Flag "%s" (reference: %s), '
+                        + "with deadline %s, has expired. "
+                        + "Consider removing this flag!",
+                        feature.name,
+                        feature.ref,
+                        feature.deadline,
+                    )
```

Comment:
Just curious why you're using % operator here, while majority of other string formatting in this update is done with f-strings?

Reply:
This is `logging` module. Thats how you use formatting for it. But I noticed one other thing. I should have used `_logger` here:)

Reply:
[TIL](https://bugs.python.org/issue46200)

ps.: mixing single and double quotes? ;)

Reply:
Yes, you got problem with that?:) Actually thats pre-commit autoformatting. I initially escaped double quotes, but it converted single quotes string with double quotes inside:)

Reply:
@dz0 also, its not just for security. logging should use its own formatting, which is that old style formatting as it actually can offer some performance gains, because it only renders placeholders when needed. But if you would put `f-string` or any other formatting, it would render all the time (even if let say different logging level is used). So I always use recommended formatting there:)